Доработана функциональность по техническому заданию 7 спринта

asdbl opened the pull request.

asdbl committed:
Добавить сохранение задач в файл в классе FileBackedTaskManager.
Добавить создание менеджера из файла в классе FileBackedTaskManager.

Yang-Pi commented on src/manager/taskManager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,203 @@
+                writer.write("\n");
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
но ты же хотел выкинуть ManagerSaveException)

Yang-Pi commented on src/manager/Managers.java:
```diff
@@ -2,14 +2,21 @@
     private Managers() {
     }
 
+    public static FileBackedTaskManager getFileBackedTaskManager() {
```
отдельный метод для нового типа менеджера не нужен, достаточно возвращать новую реализацию в существующем методе getDefault

Yang-Pi commented on src/manager/taskManager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,203 @@
+                        fileBackedTaskManager.addWithoutSave((Epic) (fromString(line)));
+                    } else if (fromString(line) instanceof Subtask) {
+                        fileBackedTaskManager.addWithoutSave((Subtask) fromString(line));
+                    } else fileBackedTaskManager.addWithoutSave(fromString(line));
```
во всех ветвлениях нужны фигурные скобки - чтобы весь код читался однообразно

Yang-Pi commented on src/manager/taskManager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,203 @@
+            while (line != null) {
+                line = bufferedReader.readLine();
+                if (line != null) {
+                    if (fromString(line) instanceof Epic) {
```
проверки на тип уже были сделаны в методе fromString , лучше передать в него fileBackedTaskManager и в соответствующей ветке case вызывать нужный метод addWithoutSave - тогда здесь не нужны будут дополнительные методы instanceof. Единственное, возможно метод нужно будет назвать по-другому, т.к. он теперь не только парсит из строки, но и добавляет таску в менеджер.

И еще один важный момент - в текущей реализации ты первый раз вызываешь fromString для проверки instanceof, а потом еще раз вызываешь уже для добавления в менеджер. Это излишне, тем более метод fromString - довольно сложный в плане выполнения. Лучше в таких случаях всегда сохранять результат в переменной, чтобы вызов тяжелого метода происходил только 1 раз

Yang-Pi commented on test/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Test
+    public void savingToFileTest() throws IOException {
```
тесты супер, покрыл много случаев)

Yang-Pi reviewed (commented):
работа отличная, есть несколько рекомендаций

asdbl committed:
Изменить метод fromString на addToManagerFromString в классе FileBackedTaskManager, теперь он сразу добавляет задачу в менеджер.
Удалить проверку на типы из метода loadFromFile в классе FileBackedTaskManager.

Yang-Pi commented on src/manager/taskManager/FileBackedTaskManager.java:
```diff
@@ -176,6 +175,8 @@ public void save() throws ManagerSaveException {
                 writer.write(toString(task));
                 writer.write("\n");
             }
+        } catch (ManagerSaveException e) {
```
не совсем так. Смотри, у тебя ведь не будет выкидываться ManagerSaveException в этом методе нигде - следовательно не получится его поймать и обработать. А вот IOException вполне может выброситься. Поэтому тебе нужно ловить IOException, а выкидывать свое кастомное ManagerSaveException - вместо RuntimeException. То есть ManagerSaveException - это просто обертка над IOException, которое станет причиной (cause)

Yang-Pi reviewed (commented):
почти)

asdbl committed:
Изменить выбрасываемое исключение в методе save в классе FileBackedTaskManager.
Удалить переопределение методов, у которых не меняется функционал в классе FileBackedTaskManager.

asdbl committed:
Изменить метод getDefault в классе Managers.
Добавить сохранение в файл при удалении задачи в классе FileBackedTaskManager.
Добавить константу HEADER для записи заголовка в файл в классе FileBackedTaskManager.

Yang-Pi commented on src/manager/taskManager/FileBackedTaskManager.java:
```diff
@@ -169,16 +137,14 @@ public static void addToManagerFromString(String value, FileBackedTaskManager fi
-            throw new ManagerSaveException("Ошибка сохранения файла.", e);
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            throw new ManagerSaveException("Ошибка сохранения в файл: " + filename, e);
```
отлично)

Yang-Pi commented on src/Main.java:
```diff
@@ -1,20 +1,17 @@
-        Path path = Paths.get("savedTasks.csv");
         InMemoryTaskManager inMemoryTaskManager = new InMemoryTaskManager();
-        FileBackedTaskManager fileBackedTaskManager = new FileBackedTaskManager(path);
+        FileBackedTaskManager fileBackedTaskManager = (FileBackedTaskManager) Managers.getDefault();
```
тебе не обязательно кастить к дочернему классу, так как у дочернего класса и интерфейса одинаковый контракт, то есть набор методов. И ты можешь обращаться к этим методам через тип интерфейса. И в целом в программировании лучше зависеть не от реализаций (конкретных классов), а от абстракций (например интерфейсов). Это важно, потому что грубо говоря тип переменной всегда остается один и тот же, и при необходимости подменить реализацию всегда остается возможным. То есть если у тебя есть taskManager переменная, объявленная с типом FileBackedTaskManager, то она может являться только объектом с типом FileBackedTaskManager и ниже, а если у нее будет тип TaskManager, то она может содержать в себе объекты любых классов, имплементирующих интерфейс

более подробно и обширно: https://alexkosarev.name/2023/05/23/solid-in-details-dip/

Yang-Pi reviewed (commented):
одна небольшая правка осталась

asdbl committed:
Изменить тип данных при создании обьекта менеджера с конкретной реализации менеджера на интерфейс TaskManager.
